database/database.rs

```diff
@@ -447,6 +459,22 @@ impl Database<WALClient> {
     }
 }
 
+fn make_checkpoint_fn(
+    path: PathBuf,
+    mut prev_checkpoint: SequenceNumber,
+    storage: Arc<MVCCStorage<WALClient>>,
+) -> impl FnMut() {
+    move || {
+        let watermark = storage.snapshot_watermark();
+        if prev_checkpoint < watermark {
+            let checkpoint = Checkpoint::new(&path).unwrap();
+            storage.checkpoint(&checkpoint).unwrap();
+            checkpoint.finish().unwrap();
+            prev_checkpoint = watermark;
+        }
```

Comment:
This can result in tearing if the checkpoint is created _right as_ data is written to the DB, though unlikely. Ideally, `Checkpoint` would lock the storage for the duration.